### PR TITLE
BREAKING CHANGE - CORE 'Add Azure Managed Identity Support to AzDO'

### PR DESCRIPTION
**Pull request by ZanattaMichael**
#### Pull Request (PR) description

The Pull Request introduces Azure DevOps Managed Identity by enabling the module to retrieve an AzureDevops API Token. This eliminates the need for a PAT. The feature is only available within Invoke-DSCResource and can be used within an Azure VM. The module achieves this by querying GET 'http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https://management.azure.com/' HTTP/1.1 Metadata: true.

> Please note that this module minimum version is PowerShell 7.

#### Task list

- [x] Added an entry to the change log under the Unreleased section of the
      file CHANGELOG.md. Entry should say what was changed and how that
      affects users (if applicable), and reference the issue being resolved
      (if applicable).
- [ ] Resource documentation updated in the resource's README.md.
- [ ] Resource parameter descriptions updated in schema.mof.
- [x] Comment-based help updated, including parameter descriptions.
- [ ] Localization strings updated.
- [x] Examples updated.
- [X] Unit tests updated. See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [ ] Integration tests updated (where possible). See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] Code changes adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

- - -
This change is [Reviewable](https://reviewable.io/reviews/dsccommunity/AzureDevOpsDsc/34)

### CODE REVIEW

File: tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1 [hunk 1]
```diff
@@ -3,20 +3,43 @@
         Automated unit test for classes in AzureDevOpsDsc.
 #>
 
-Import-Module -Name (Join-Path -Path $PSScriptRoot -ChildPath '\..\Modules\TestHelpers\CommonTestHelper.psm1')
-Import-Module -Name (Join-Path -Path $PSScriptRoot -ChildPath '\..\Modules\TestHelpers\CommonTestCases.psm1')
+
+Function Split-RecurivePath {
```
Comment (kilasuit) [inline tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1]:
```suggestion
Function Split-RecursivePath {
```

File: tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1 [hunk 2]
```diff
@@ -3,20 +3,43 @@
...
 $script:dscModuleName = 'AzureDevOpsDsc'
 $script:dscModule = Get-Module -Name $script:dscModuleName -ListAvailable | Select-Object -First 1
 $script:dscModuleFile = $($script:dscModule.ModuleBase +'\'+ $script:dscModuleName + ".psd1")
+
 Get-Module -Name $script:dscModuleName -All |
     Remove-Module $script:dscModuleName -Force -ErrorAction SilentlyContinue
 
 $script:subModuleName = 'AzureDevOpsDsc.Common'
-Import-Module -Name $script:dscModuleFile -Force
+Import-Module -Name $script:dscModuleFile #-Force
```
Comment (kilasuit) [inline tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1]: What was the reason in commenting out the `-Force`

File: tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1 [hunk 3]
```diff
@@ -3,20 +3,43 @@
...
 
 Get-Module -Name $script:subModuleName -All |
     Remove-Module -Force -ErrorAction SilentlyContinue
 $script:subModulesFolder = Join-Path -Path $script:dscModule.ModuleBase -ChildPath 'Modules'
 $script:subModuleFile = Join-Path $script:subModulesFolder "$($script:subModuleName)/$($script:subModuleName).psd1"
-Import-Module -Name $script:subModuleFile -Force #-Verbose
+Import-Module -Name $script:subModuleFile #-Force #-Verbose
```
Comment (kilasuit) [inline tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1]: What was the reason in commenting out the `-Force` & `-Verbose`

**Review by kilasuit — CHANGES_REQUESTED** [review]

File: tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1 (hunk 3 above)
Comment (ZanattaMichael) [reply tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1]: Testing. I forgot to remove it.

File: tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1 (hunk 2 above)
Comment (ZanattaMichael) [reply tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1]: Testing. I forgot to remove it.

File: tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1 [hunk 4]
```diff
@@ -3,20 +3,43 @@
...
+
+    1 .. $Times | ForEach-Object {
+        $Path = Split-Path -Path $Path -Parent
+    }
+
+    $Path
+}
+
+
+$script:RepositoryRoot = Split-RecurivePath $PSScriptRoot -Times 4
```
Comment (kilasuit) [inline tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1]:
```suggestion
$script:RepositoryRoot = Split-RecursivePath $PSScriptRoot -Times 4
```

**Review by kilasuit — CHANGES_REQUESTED** [review]

File: tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1 (hunk 4 above)
Comment (kilasuit) [reply tests/Unit/Modules/AzureDevOpsDsc.Common/AzureDevOpsDsc.Common.TestInitialization.ps1]: Sorry I missed this last time too

**Review by kilasuit — DISMISSED** [review]

**Comment by ZanattaMichael** [comment]
@kilasuit So the module minimum version will be version 7. Is that going to be an issue? I figured that it wouldn't since the module resources are incomplete and it's easier to write for the new version.

**Comment by kilasuit** [comment]
Sorry on the delay in getting back to you on this, I know it's been a while.

> @kilasuit So the module minimum version will be version 7. Is that going to be an issue? I figured that it wouldn't since the module resources are incomplete and it's easier to write for the new version.

As you'll have seen in the DSC Channel in the PowerShell Discord/slack DSC I've asked others for input on this generally with modules part of the DSCCommunity

Personally I'm ok with it, if we do can potentially have releases that are both v5 and v7 compatible (for those orgs not upgrading to v7 yet) side by side & to do that we'd need to
- Bump the major version of the module with this release to show that v7 language features are being used
- merge this into a v7 branch not the main branch so can leave main as is for now for v5 compatible changes
- update docs about this 

All of the above I've mentioned other is maintainer work, so let me get that done, if that is what others agree as a way forward.

File: source/Classes/004.ManagedIdentityToken.ps1 [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        # If all checks pass, return true
+        Write-Verbose "[ManagedIdentityToken] The hashtable is valid and contains all the expected keys."
+        return $true
+    }
+
+    # Function to convert a SecureString to a String
+    hidden [string]ConvertFromSecureString([SecureString]$SecureString) {
+        # Convert a SecureString to a String
+        $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($SecureString)
+        $String = ($this.linux) ? [System.Runtime.InteropServices.Marshal]::PtrToStringUni($BSTR) :
```
Comment (kilasuit) [inline source/Classes/004.ManagedIdentityToken.ps1:53]: based on discussion can we change this to be able to back to v5 syntax and check based on discussion in this review

File: source/AzureDevOpsDsc.psd1 [hunk 6]
```diff
@@ -20,7 +20,7 @@
     Description        = 'Module with DSC Resources for deployment and configuration of Azure DevOps Server/Services.'
 
     # Minimum version of the Windows PowerShell engine required by this module
-    PowerShellVersion  = '5.0'
+    PowerShellVersion  = '7.0'
```
Comment (kilasuit) [inline source/AzureDevOpsDsc.psd1:23]:
```suggestion
    PowerShellVersion  = '5.1'
```

File: source/Classes/004.ManagedIdentityToken.ps1 [hunk 7]
```diff
@@ -0,0 +1,120 @@
+
+
+class ManagedIdentityToken {
+
+    [SecureString]$access_token
+    [DateTime]$expires_on
+    [int]$expires_in
+    [string]$resource
+    [string]$token_type
+    hidden [bool]$linux = $IsLinux
```
Comment (kilasuit) [inline source/Classes/004.ManagedIdentityToken.ps1:10]: Perhaps this shouldn't be a check for 1 OS type unless this is definitely an only linux and not also MacOS issue to check for & not use the `$Is` automatic varibales for checking and use the v5 compatable `[Environment]::OSVersion.Platform` instead

**Review by kilasuit — CHANGES_REQUESTED** [review]
Can we change these bits in this PR, ideally in a single commit so can in future look at v7 compatible codebases which is a larger longer term discussion point for DSCCommunity to have

Also no need for bumping Major version in this PR or merging into a v7 branch.
We can do those later if we decide to use v7 compatible syntax

**Comment by kilasuit** [comment]
The previous comment was made based on the comment made not from having relooked at the code in this PR as was an interesting discussion point.

However on the disccusion in the DSC Channel in the slack/discord and investigation on how to check for compatibility I'm going to ask for simplicity for maintainence and use going forward for the following

- Revert to using 5.1 language features as part of this PR so can get this merged sooner rather than later & users can look to use this in their configurations.
- As part of longer term plans, bring this discussion point to the next DSC Community call, and raise a discussion for this in the PowerShell discussions org as we in DSCCommunity don't currently have discussions enabled 
- Based on the above discussion, also discuss potentially using this module as a basis for a POC of maintaining v5 and v7 together using the method I mentioned in previous comment
  - This is cleaner way forward than forking this into yet another repo that needs managed and maintained and then releasing a `AzureDevOpsDscV7` module instead (which @johlju suggested in the discussion in the dsc channel) which from historics leads to causing much more more confusion than needed when we can totally avoid that using Major Versions for seperation which I feel is the better way, if not the most correct way to do this, especially as we should keep alignment to Semver and can be easily incorporated into changelogs too which again aids discoverability

**Comment by ZanattaMichael** [comment]
> The previous comment was made based on the comment made not from having relooked at the code in this PR as was an interesting discussion point.
> 
> However on the disccusion in the DSC Channel in the slack/discord and investigation on how to check for compatibility I'm going to ask for simplicity for maintainence and use going forward for the following
> 
> * Revert to using 5.1 language features as part of this PR so can get this merged sooner rather than later & users can look to use this in their configurations.
> * As part of longer term plans, bring this discussion point to the next DSC Community call, and raise a discussion for this in the PowerShell discussions org as we in DSCCommunity don't currently have discussions enabled
> * Based on the above discussion, also discuss potentially using this module as a basis for a POC of maintaining v5 and v7 together using the method I mentioned in previous comment
>   
>   * This is cleaner way forward than forking this into yet another repo that needs managed and maintained and then releasing a `AzureDevOpsDscV7` module instead (which @johlju suggested in the discussion in the dsc channel) which from historics leads to causing much more more confusion than needed when we can totally avoid that using Major Versions for seperation which I feel is the better way, if not the most correct way to do this, especially as we should keep alignment to Semver and can be easily incorporated into changelogs too which again aids discoverability

No worries. My reasoning for pwsh 7.0 to force the use of Invoke-DSC Resource rather then using the MOF since this configuration is best run within a pipeline so pwsh 7.0 would be preferred. I need to sit down and do some more work on this module, since I've been writing resources for work and I needed to refactor some of the MI code. I also plan to abstract out the PAT token, and possibly add client ID/ secret support. This will require some major rework of the module.

Thanks for getting back to me. Sorry I would of responded on discord, however I've been busy with #DadLife.

Cheers,

Michael.

**Comment by ZanattaMichael** [comment]
Closing PR due to new branch with additional features.

**Pull request closed without merging**